dali/util/shared_mem.cc

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <fcntl.h> /* For O_* constants */
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <exception>
+#include <cstring>
+
+#include "dali/util/shared_mem.h"
+
+namespace dali {
+
+namespace detail {
+
+bool dir_exists(const char * str) {
+  struct stat sb;
+  if (stat(str, &sb) == -1) {
+    return false;
+  }
+  return S_ISDIR(sb.st_mode);
+}
+
+}  // namespace detail
+
+namespace python {
+ShmFdWrapper::ShmFdWrapper(int fd) : fd_{fd} {}
+
+ShmFdWrapper::ShmFdWrapper() {
+  // Abstract away the fact that shm_open requires filename.
+  constexpr char dev_shm_path[] = "/dev/shm/";
+  constexpr char run_shm_path[] = "/run/shm/";
+  constexpr char temp_filename_template[] = "nvidia_dali_XXXXXX";
+  constexpr int kShmPathSize = sizeof(dev_shm_path) - 1;
+  constexpr int kPathSize = kShmPathSize + sizeof(temp_filename_template);
+  char temp_path[kPathSize];
+  const char *shm_path = nullptr;
+  if (detail::dir_exists(dev_shm_path)) {
+    shm_path = dev_shm_path;
+  } else if (detail::dir_exists(run_shm_path)) {
+    shm_path = run_shm_path;
+  } else {
+    throw std::runtime_error("shared memory dir not found");
+  }
+  memcpy(temp_path, shm_path, kShmPathSize);
+  memcpy(temp_path + kShmPathSize, temp_filename_template, sizeof(temp_filename_template));
+  int temp_fd = mkstemp(temp_path);
+  if (temp_fd < 0) {
+    throw std::runtime_error("temporary file creation failed");
+  }
+  if (unlink(temp_path) != 0) {
+    throw std::runtime_error("couldn't unlink temporary file");
+  }
+  const char * temp_filename = temp_path + kShmPathSize;
+  fd_ = shm_open(temp_filename, O_CREAT | O_EXCL | O_RDWR, S_IRUSR | S_IWUSR);
+  if (fd_ < 0) {
+    throw std::runtime_error("shm_open call failed");
+  }
+  ::close(temp_fd);
```

Comment:
Why we cannot close just after the unlink?

Reply:
To keep the file alive till it is used in the shm_open.